util/topgen/merge.py

```diff
@@ -119,6 +124,13 @@ def elaborate_instance(instance, block: IpBlock):
 
     instance["param_list"] = new_params
 
+    # for each module declaration, check to see that the parameter actually exists
+    # and can be set
+    for decl in instance["param_decl"].keys():
+        if not any(decl in param['name'] for param in instance["param_list"]):
+            log.error("{} is not a valid parameter of {} that can be "
+                      "set from top level".format(decl, block.name))
```

Comment:
how about we directly override the `default` values in this function? by doing that, we could do away with the additional helper function in the template (I've done something similar with the `MemSize` parameter in #7267). 

for example, we could add the following `elif` leg to the `if isinstance(param, RandParameter):` statement on line 109 above:

```python
        # if this exposed parameter is listed in the `param_decl` dict, 
        # override its default value.
        elif param.name in instance["param_decl"]:
            new_param['default'] = instance["param_decl"][param.name]
            # remove the parameter from the accounting dict
            del param_decl_accounting[param.name]
```

`param_decl_accounting` would have to be initialized with 
```python
param_decl_accounting = instance["param_decl"].copy()
```
before the loop on line 77. this accounting dict can then be used to check that all parameters have been matched after the loop on line 126
```python
    # we should not have left-over declarations here.
    for decl in param_decl:
         log.error("{} is not a valid parameter of {} that can be "
                   "set from top level".format(decl, block.name))
```

Reply:
yeah i thought about overriding also, but it felt a bit like contaminating the definition of "default" by doing that. 
I don't have strong feelings one way or the other though, so if others also feel like overriding is the better option, I'm happy to do that.

Reply:
Yeah I know, `default` feels a bit like a misnomer in this case... I think it has originally been named `default` since this is the value that is assigned in the `top_earlgrey` module parameter definition, which can later be overridden in RTL wherever the `top_earlgrey` module is instantiated.

Given that we already override these values here for the `RndCnst` parameters, shall we just keep the override mechanism here to make this a bit more contained?

Reply:
sorry when you say keep it here, you mean override in merge and not at top level? 
I'm at moment inclined towards  top level out of...laziness :D

Reply:
after looking at this more closely, i think you are right. It is better to change it here since the `Rand` case already does the same.
I've done mostly what you suggested, with a slight change (didn't use copy..since i can't quite tell if this puts some of the underlying objects at risk if we added them in the future).